rainier-core/src/main/scala/com/stripe/rainier/compute/LogLineOps.scala

```diff
@@ -24,6 +24,58 @@ private[compute] object LogLineOps {
    */
   def log(@unused line: LogLine): Option[Real] = None
 
+  /*
+  If possible, return a representation of ax as a list of terms to be summed. Within limits, it's helpful here
+  to distribute the multiplications across any internal sums, because that may well surface some terms
+  that are in common with whatever we're about to sum this with.
+   */
+  val DistributeToMaxTerms = 20
+  def distribute(line: LogLine): Option[Line] = {
+
+    def nTerms(l: Line): Int =
+      if (l.b == 0.0)
+        l.ax.size
+      else
+        l.ax.size + 1
+    def nTerms2(l: Line): Int = {
+      val n = nTerms(l)
+      (n * (n + 1)) / 2
+    }
+
+    val initial = (List.empty[(NonConstant, Double)], Option.empty[Line])
+    val (factors, terms) = line.ax.foldLeft(initial) {
```

Comment:
have you considered using recursion instead of foldLeft? i think it would be easier to read; especially the condition when you bail out for optimizations would be more direct

Reply:
I agree that this is kinda hard to read but I'm not sure how recursion would help, it's structurally a pretty straightforward foldLeft. (Can you explain what you mean by "bail out"?Which line?)

Reply:
line 57, i thought this was a fall-through condition that, once reached, will be exercised for the rest of foldLeft but i think i read this wrong
you could hit that pattern and then still hit other cases for subsequent values of ax, right?

Reply:
that's right.